src/library_openal.js

```diff
@@ -1,1622 +1,4591 @@
+      var posY = src.position[1];
+      var posZ = src.position[2];
+      var dirX = src.direction[0];
+      var dirY = src.direction[1];
+      var dirZ = src.direction[2];
 
-  alcGetIntegerv: function(device, param, size, data) {
-    if (size == 0 || !data) {
-      AL.currentContext.err = 0xA003 /* AL_INVALID_VALUE */;
-      return;
-    }
+      var listener = src.context.audioCtx.listener;
+      var lPosX = listener._position[0];
+      var lPosY = listener._position[1];
+      var lPosZ = listener._position[2];
 
-    switch(param) {
-    case 0x1000 /* ALC_MAJOR_VERSION */:
-      {{{ makeSetValue('data', '0', '1', 'i32') }}};
-      break;
-    case 0x1001 /* ALC_MINOR_VERSION */:
-      {{{ makeSetValue('data', '0', '1', 'i32') }}};
-      break;
-    case 0x1002 /* ALC_ATTRIBUTES_SIZE */:
-      if (!device) {
-        AL.alcErr = 0xA001 /* ALC_INVALID_DEVICE */;
-        return 0;
+      // WebAudio expects world space coordinates, so if the source is listener-relative
+      // we must transform the coordinates from listener space into world space.
+      if (src.relative) {
+        // Negate the listener direction since forward is -Z.
+        var lBackX = -listener._direction[0];
+        var lBackY = -listener._direction[1];
+        var lBackZ = -listener._direction[2];
+        var lUpX = listener._up[0];
+        var lUpY = listener._up[1];
+        var lUpZ = listener._up[2];
+
+        // Normalize the Back vector
+        var invMag = 1.0 / Math.sqrt(lBackX * lBackX + lBackY * lBackY + lBackZ * lBackZ);
+        lBackX *= invMag;
+        lBackY *= invMag;
+        lBackZ *= invMag;
+
+        // ...and the Up vector
+        var invMag = 1.0 / Math.sqrt(lUpX * lUpX + lUpY * lUpY + lUpZ * lUpZ);
+        lUpX *= invMag;
+        lUpY *= invMag;
+        lUpZ *= invMag;
+
+        // Calculate the Right vector as the cross product of the Up and Back vectors
+        var lRightX = (lUpY * lBackZ - lUpZ * lBackY);
+        var lRightY = (lUpZ * lBackX - lUpX * lBackZ);
+        var lRightZ = (lUpX * lBackY - lUpY * lBackX);
+
+        var oldX = dirX;
+        var oldY = dirY;
+        var oldZ = dirZ;
+
+        // Use our 3 vectors to apply a change-of-basis matrix to the source direction
+        dirX = oldX * lRightX + oldY * lUpX + oldZ * lBackX;
+        dirY = oldX * lRightY + oldY * lUpY + oldZ * lBackY;
+        dirZ = oldX * lRightZ + oldY * lUpZ + oldZ * lBackZ;
+
+        var oldX = posX;
+        var oldY = posY;
+        var oldZ = posZ;
+
+        // ...and to the source position
+        posX = oldX * lRightX + oldY * lUpX + oldZ * lBackX;
+        posY = oldX * lRightY + oldY * lUpY + oldZ * lBackY;
+        posZ = oldX * lRightZ + oldY * lUpZ + oldZ * lBackZ;
+
+        // The change-of-basis corrects the orientation, but the origin is still the listener.
+        // Translate the source position by the listener position to finish.
+        posX += lPosX;
+        posY += lPosY;
+        posZ += lPosZ;
```

Comment:
In listener relative mode, the source position represents a displacement from the listener, expressed in a coordinate system where the listener's forward vector is -Z. The change-of-basis corrects the displacements so they're expressed in world-space instead of listener-space. This is enough for the source direction, since the direction is always relative to the origin anyway, but since the position is a displacement from the listener, we translate it by the listener position to get the final world space coordinates.

I've added a much more detailed comment to explain what's going on here. I'm pretty sure the math is correct, but perhaps more rigorous testing of more complex cases is the only way to be sure.